From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Marijn Suijten <marijn.suijten@somainline.org>
Date: Mon, 8 Jul 2024 22:56:46 +0200
Subject: [PATCH] TimeKeep: Replace `SystemProperties` private API with
 `sysprop` definition

Using `android.os.SystemProperties` requires access to Android
private platform APIs which is incompatible with specifying an SDK
version.  Opting in to private APIs is largely unnecessary when the
_vendor-specific_ system property can be accessed through a generated
`sysprop` library.

Unfortunately these libraries are not available to C code, only to
C++, so we'll keep using the hardcoded property name in `timekeep.c`.
---
 Android.mk                             |  3 ++-
 src/com/sony/timekeep/TimeKeep.java    |  8 +++-----
 sysprop/Android.bp                     |  8 ++++++++
 sysprop/SonyTimekeepProperties.sysprop | 10 ++++++++++
 4 files changed, 23 insertions(+), 6 deletions(-)
 create mode 100644 sysprop/Android.bp
 create mode 100644 sysprop/SonyTimekeepProperties.sysprop

diff --git a/Android.mk b/Android.mk
index 9af9d743aa38b71a8e1b073417d81b73ec9db464..41050c87b6cd1e83ac20e694849091f5b89177a1 100644
--- a/Android.mk
+++ b/Android.mk
@@ -19,8 +19,9 @@ include $(CLEAR_VARS)
 LOCAL_MODULE_TAGS := optional
 LOCAL_SRC_FILES := $(call all-java-files-under, src)
 LOCAL_PACKAGE_NAME := TimeKeep
+LOCAL_STATIC_JAVA_LIBRARIES := SonyTimekeepProperties
 LOCAL_CERTIFICATE := platform
-LOCAL_PRIVATE_PLATFORM_APIS := true
+LOCAL_SDK_VERSION := current
 LOCAL_PRIVILEGED_MODULE := true
 LOCAL_PROGUARD_ENABLED := disabled
 ifneq ($(call math_gt_or_eq, $(PLATFORM_SDK_VERSION), 28),)
diff --git a/src/com/sony/timekeep/TimeKeep.java b/src/com/sony/timekeep/TimeKeep.java
index 91b3e3b48d0f68a8724a4df82334082ceb2a0097..a1300808d463ce3b32f2a4a0c47ffb2ec773e0f7 100644
--- a/src/com/sony/timekeep/TimeKeep.java
+++ b/src/com/sony/timekeep/TimeKeep.java
@@ -46,11 +46,11 @@ import android.content.BroadcastReceiver;
 import android.content.Context;
 import android.content.Intent;
 import android.util.Log;
-import android.os.SystemProperties;
+
+import com.sony.timekeep.SonyTimekeepProperties;
 
 public class TimeKeep extends BroadcastReceiver {
 	private static final String TAG = "TimeKeep-Receiver";
-	private static final String TIMEADJ_PROP = "persist.vendor.timeadjust";
 	private static final String RTC_SINCE_EPOCH = "/sys/class/rtc/rtc0/since_epoch";
 	private static final String RTC_ATS_FILE = android.os.Build.VERSION.SDK_INT >= 26 ?
 		"/data/vendor/time/ats_2" : "/data/time/ats_2";
@@ -63,10 +63,8 @@ public class TimeKeep extends BroadcastReceiver {
 		long epoch_since = readEpoch();
 		seconds -= epoch_since;
 
-		String currentAdjust = SystemProperties.get(TIMEADJ_PROP);
-
 		Log.d(TAG, "Setting adjust property to " + seconds);
-		SystemProperties.set(TIMEADJ_PROP, Long.toString(seconds));
+		SonyTimekeepProperties.timeadjust(Long.toString(seconds));
 		writeATS(seconds);
 	}
 
diff --git a/sysprop/Android.bp b/sysprop/Android.bp
new file mode 100644
index 0000000000000000000000000000000000000000..09142c67da95c7a36f5a2ef9dc3e8e9174e62fe0
--- /dev/null
+++ b/sysprop/Android.bp
@@ -0,0 +1,8 @@
+sysprop_library {
+    name: "SonyTimekeepProperties",
+    proprietary: true,
+
+    srcs: ["SonyTimekeepProperties.sysprop"],
+
+    property_owner: "Vendor",
+}
diff --git a/sysprop/SonyTimekeepProperties.sysprop b/sysprop/SonyTimekeepProperties.sysprop
new file mode 100644
index 0000000000000000000000000000000000000000..b4023af6a4d94c0274d7e71b019756929fc89deb
--- /dev/null
+++ b/sysprop/SonyTimekeepProperties.sysprop
@@ -0,0 +1,10 @@
+module: "com.sony.timekeep.SonyTimekeepProperties"
+owner: Vendor
+
+prop {
+    api_name: "timeadjust"
+    type: String
+    scope: Internal
+    access: ReadWrite
+    prop_name: "persist.vendor.timeadjust"
+}
